app_dart/lib/src/foundation/utils.dart

```diff
@@ -29,48 +32,71 @@ Duration twoSecondLinearBackoff(int attempt) {
+  return retryOptions.retry(
+    () async => await getUrl(url, httpClientProvider, log, timeout: timeout),
+    retryIf: (Exception e) => e is HttpException,
+  );
+}
 
-  final HttpClient client = branchHttpClientProvider();
+/// Return [String] of response from [url] if status is [HttpStatus.ok].
+///
+/// If [url] returns [HttpStatus.notFound] throw [NotFoundException].
+/// Otherwise, throws [HttpException].
+FutureOr<String> getUrl(
+  Uri url,
+  HttpClientProvider httpClientProvider,
+  Logging log, {
+  Duration timeout = const Duration(seconds: 5),
+}) async {
+  final HttpClient client = httpClientProvider();
   try {
-    // TODO(keyonghan): apply retry logic here to simply, https://github.com/flutter/flutter/issues/52427
-    for (int attempt = 0; attempt < 3; attempt++) {
-      try {
-        final HttpClientRequest clientRequest = await client.getUrl(url).timeout(timeout);
-        final HttpClientResponse clientResponse = await clientRequest.close().timeout(timeout);
-        final int status = clientResponse.statusCode;
-
-        if (status == HttpStatus.ok) {
-          final String content = await utf8.decoder.bind(clientResponse).join();
-          return content;
-        } else {
-          log.warning('Attempt to download $filePath failed (HTTP $status)');
-        }
-      } catch (error, stackTrace) {
-        log.error('Attempt to download $filePath failed:\n$error\n$stackTrace');
-      }
-      await Future<void>.delayed(gitHubBackoffCalculator(attempt));
+    final HttpClientRequest clientRequest = await client.getUrl(url).timeout(timeout);
+    final HttpClientResponse clientResponse = await clientRequest.close().timeout(timeout);
+    final int status = clientResponse.statusCode;
+
+    if (status == HttpStatus.ok) {
+      return await utf8.decoder.bind(clientResponse).join();
+    } else if (status == HttpStatus.notFound) {
+      throw NotFoundException('HTTP $status: $url');
+    } else {
+      log.warning('HTTP $status: $url');
+      throw HttpException('HTTP $status: $url');
     }
   } finally {
-    client.close(force: true);
+    client.close();
   }
-  log.error('GitHub not responding; giving up');
-  return null;
 }
 
 /// Gets supported branch list of `flutter/flutter` via GitHub http request.
 Future<Uint8List> getBranches(
-    HttpClientProvider branchHttpClientProvider, Logging log, GitHubBackoffCalculator gitHubBackoffCalculator) async {
-  String content = await remoteFileContent(
-      branchHttpClientProvider, log, gitHubBackoffCalculator, '/flutter/cocoon/master/app_dart/dev/branches.txt');
-  content ??= 'master';
+  HttpClientProvider httpClientProvider,
+  Logging log, {
+  RetryOptions retryOptions,
+}) async {
+  String content;
+  try {
+    content = await githubFileContent(
+      '/flutter/cocoon/master/app_dart/dev/branches.txt',
+      httpClientProvider: httpClientProvider,
+      log: log,
+      retryOptions: retryOptions,
+    );
+  } on HttpException {
```

Comment:
In what situation would this happen? And should we trigger some sort of alert if we hit this?

Reply:
This would only happen during a GitHub CDN outage (which we haven't faced issues with). The implementers of this function don't catch this error, so it would surface as an increased error rate in cocoon (we already use error tracking dashboards)

Reply:
If `NotFoundException`, we should create an empty config. This case is expected when we remove the existing configs (devicelab/manifest.yaml, *_builders.json) or when a release branch is created without .ci.yaml.